### PR TITLE
[branch-2.1](binlog) Support drop view binlog (#39781) (#33321)

### PR DESCRIPTION
**Pull request by wyxxxcat**
pick #39781
changes of #39781  depend on deleteIndexList of #33321

### CODE REVIEW

**Comment by wyxxxcat** [comment]
run buildall

**Comment by wyxxxcat** [comment]
run buildall

**Comment by wyxxxcat** [comment]
run buildall

**Comment by wyxxxcat** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/datasource/InternalCatalog.java [hunk 1]
```diff
@@ -1044,6 +1044,42 @@ public void replayRecoverTable(RecoverInfo info) throws MetaNotFoundException, D
...
+        AgentTaskExecutor.submit(batchTask);
+    }
+
+    public void erasePartitionDropBackendReplicas(List<Partition> partitions) {
+        // no need send be delete task, when be report its tablets, fe will send delete task then.
+    }
+
+    public void eraseDroppedIndexBackendReplicas(long tableId, List<Long> indexIdList) {
+        // nothing to do in non cloud mode
+    }
```
Comment (w41ter) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/InternalCatalog.java:1081]: Remove it

File: fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java
Comment (w41ter) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java:1]: Pick this file is unnecessary

File: fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java [hunk 2]
```diff
@@ -2094,29 +2094,24 @@ private Status allTabletCommitted(boolean isReplay) {
...
+                }
+                if (tableType == TableType.OLAP && backupTableInfo != null) {
+                    // drop the non restored partitions.
+                    dropNonRestoredPartitions(db, (OlapTable) table, backupTableInfo);
+                } else if (isCleanTables) {
+                    // otherwise drop the entire table.
+                    LOG.info("drop non restored table {}({}). {}", tableName, tableId, this);
+                    boolean isView = false;
+                    boolean isForceDrop = false; // move this table into recyclebin.
+                    env.getInternalCatalog().dropTableWithoutCheck(db, table, isView, isForceDrop);
```
Comment (w41ter) [inline fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java:2114]: @wyxxxcat This patch is staled, please update.

File: fe/fe-core/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java [hunk 3]
```diff
@@ -1049,6 +1056,10 @@ public void replayDropRollup(DropInfo dropInfo, Env env) throws MetaNotFoundExce
         } finally {
             olapTable.writeUnlock();
         }
+
+        List<Long> deleteIndexList = new ArrayList<Long>();
+        deleteIndexList.add(rollupIndexId);
+        Env.getCurrentInternalCatalog().eraseDroppedIndexBackendReplicas(olapTable.getId(), deleteIndexList);
```
Comment (w41ter) [inline fe/fe-core/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:1062]: remove it

File: fe/fe-core/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java [hunk 4]
```diff
@@ -962,6 +968,7 @@ public void processDropMaterializedView(DropMaterializedViewStmt dropMaterialize
         } finally {
             olapTable.writeUnlock();
         }
+        Env.getCurrentInternalCatalog().eraseDroppedIndexBackendReplicas(olapTable.getId(), deleteIndexList);
```
Comment (w41ter) [inline fe/fe-core/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:971]: remove it

File: fe/fe-core/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java [hunk 5]
```diff
@@ -951,7 +956,8 @@ public void processDropMaterializedView(DropMaterializedViewStmt dropMaterialize
             // Step3: log drop mv operation
             EditLog editLog = Env.getCurrentEnv().getEditLog();
             editLog.logDropRollup(
-                    new DropInfo(db.getId(), olapTable.getId(), olapTable.getName(), mvIndexId, false, 0));
+                    new DropInfo(db.getId(), olapTable.getId(), olapTable.getName(), mvIndexId, false, false, 0));
+            deleteIndexList.add(mvIndexId);
```
Comment (w41ter) [inline fe/fe-core/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:960]: remove

File: fe/fe-core/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java [hunk 6]
```diff
@@ -931,15 +933,18 @@ public void processBatchDropRollup(List<AlterClause> dropRollupClauses, Database
...
                     String.join("", rollupNameSet), olapTable.getName());
         } finally {
             olapTable.writeUnlock();
         }
+        Env.getCurrentInternalCatalog().eraseDroppedIndexBackendReplicas(olapTable.getId(), deleteIndexList);
     }
 
     public void processDropMaterializedView(DropMaterializedViewStmt dropMaterializedViewStmt, Database db,
             OlapTable olapTable) throws DdlException, MetaNotFoundException {
+        List<Long> deleteIndexList = new ArrayList<Long>();
```
Comment (w41ter) [inline fe/fe-core/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:947]: remove

File: fe/fe-core/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java [hunk 7]
```diff
@@ -931,15 +933,18 @@ public void processBatchDropRollup(List<AlterClause> dropRollupClauses, Database
             long tableId = olapTable.getId();
             String tableName = olapTable.getName();
             editLog.logBatchDropRollup(new BatchDropInfo(dbId, tableId, tableName, indexIdSet));
+            deleteIndexList = indexIdSet.stream().collect(Collectors.toList());
             LOG.info("finished drop rollup index[{}] in table[{}]",
                     String.join("", rollupNameSet), olapTable.getName());
         } finally {
             olapTable.writeUnlock();
         }
+        Env.getCurrentInternalCatalog().eraseDroppedIndexBackendReplicas(olapTable.getId(), deleteIndexList);
```
Comment (w41ter) [inline fe/fe-core/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:942]: remove

File: fe/fe-core/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java [hunk 8]
```diff
@@ -931,15 +933,18 @@ public void processBatchDropRollup(List<AlterClause> dropRollupClauses, Database
             long tableId = olapTable.getId();
             String tableName = olapTable.getName();
             editLog.logBatchDropRollup(new BatchDropInfo(dbId, tableId, tableName, indexIdSet));
+            deleteIndexList = indexIdSet.stream().collect(Collectors.toList());
```
Comment (w41ter) [inline fe/fe-core/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:936]: remove

File: fe/fe-core/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java [hunk 9]
```diff
@@ -901,6 +902,7 @@ private long checkAndGetBaseIndex(String baseIndexName, OlapTable olapTable) thr
 
     public void processBatchDropRollup(List<AlterClause> dropRollupClauses, Database db, OlapTable olapTable)
             throws DdlException, MetaNotFoundException {
+        List<Long> deleteIndexList = null;
```
Comment (w41ter) [inline fe/fe-core/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:905]: remove

**Review by w41ter — CHANGES_REQUESTED** [review]

**Comment by w41ter** [comment]
close in favor #44112

**Pull request closed without merging**